File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        vector<int> prefixMax(n);
        prefixMax[0] = height[0];
        for(int i = 1 ; i < n;i++){
            prefixMax[i] = max(prefixMax[i-1] , height[i]);
        }
        vector<int> suffixMax(n);
        suffixMax[n-1] = height[n-1];
        for(int i = n-2; i>=0; i--){
            suffixMax[i] = max(suffixMax[i+1] , height[i]);
        }
        int total = 0;
        for(int i = 0 ; i< n; i++){
            int Lmax = prefixMax[i];
            int Rmax = suffixMax[i];
            if(height[i] < Lmax && height[i] < Rmax){
                total += min(Lmax, Rmax) - height[i];
            }
        }
        return total;
    }
};

// Synced seamlessly with LeetHub Pro
// Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
// Get it here: https://chromewebstore.google.com/detail/bcilpkkbokcopmabingnndookdogmbna

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        prefix = [0] * n
        prefix[0] = height[0]
        for i in range(1,n):
            prefix[i] = max(prefix[i-1],height[i])
        sufix = [0] * n
        sufix[n-1] = height[n-1]
        for i in range(n-2,-1,-1):
            sufix[i] = max(sufix[i+1],height[i])
        
        total = 0
        for i in range(n):
            Lmax = prefix[i]
            Rmax = sufix[i]
            if height[i] < Lmax and height[i] < Rmax:
                total += min(Lmax,Rmax)-height[i]
        
        return total
        


# Synced seamlessly with LeetHub Pro
# Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
# Get it here: https://chromewebstore.google.com/detail/bcilpkkbokcopmabingnndookdogmbna